src/modules/fancyzones/lib/ZoneWindow.cpp

```diff
@@ -81,6 +81,44 @@ namespace ZoneWindowUtils
         }
         return std::wstring{ uniqueId };
     }
+
+    const CLSID CLSID_ImmersiveShell = { 0xC2F03A33, 0x21F5, 0x47FA, 0xB4, 0xBB, 0x15, 0x63, 0x62, 0xA2, 0xF2, 0x39 };
+
+    IServiceProvider* GetServiceProvider()
+    {
+        IServiceProvider* provider{ nullptr };
+        if (FAILED(CoCreateInstance(CLSID_ImmersiveShell, nullptr, CLSCTX_LOCAL_SERVER, __uuidof(provider), (PVOID*)&provider)))
+        {
+            return nullptr;
+        }
+        return provider;
+    }
+
+    IVirtualDesktopManager* GetVirtualDesktopManager()
+    {
+        IVirtualDesktopManager* manager{ nullptr };
+        IServiceProvider* serviceProvider = GetServiceProvider();
+        if (serviceProvider == nullptr || FAILED(serviceProvider->QueryService(__uuidof(manager), &manager)))
+        {
+            return nullptr;
+        }
+        return manager;
+    }
+
+    bool GetWindowDesktopId(HWND topLevelWindow, GUID* desktopId)
+    {
+        static IVirtualDesktopManager* virtualDesktopManager = GetVirtualDesktopManager();

```

Comment:
We should only initialize the service provider and the `VirtualDesktopManager` once at FZ startup.
Move this to a different file, create a singleton that initialize the method, taking in consideration it may fail, in that case simple ignore the failure and always return nullptr in `GetWindowDesktopId`

Reply:
But wouldn’t this initialize it only once as it is declared as static according to C++ standard. Also it will guarantee that initialization is thread safe.

Reply:
Sorry, it's late and I didn't see it. OK good.
Please move it outside of this file.

Reply:
Moved to `VirtualDesktopUtils.h(.cpp)`